content/blog/2019/06/2019-06-14-performance-testing-jenkins.adoc

```diff
@@ -0,0 +1,129 @@
+benchmarks from your IDE, just like unit tests. You can easily configure your benchmarks by either using your
+Java methods, or by using Jenkins Configuration-as-Code plugin and passing the path to your YAML file.
+
+=== Writing benchmarks
+
+A reference to the Jenkins instance is available through either `JmhBenchmarkState#getJenkins()` or through
+`Jenkins.getInstance()` like you would otherwise do. `JmhBenchmarkState` provides `setup()` and `tearDown` methods
+which can be overridden to configure the Jenkins instance according to your benchmark's requirements.
+
+=== Running the benchmarks
+
+The benchmarks can be run through JUnit tests. From a test method, you can use the `OptionsBuilder` provided by JMH to
+configure your benchmarks. Benchmark reports generated (in JSON) can be visualized using the
+link:https://wiki.jenkins.io/display/JENKINS/JMH+Report+Plugin[JMH Report Plugin].
+
+==== Automatic detection of benchmark classes
+
+Classes containing benchmarks are found automatically by the `BenchmarkFinder`
+when annotated with `@JmhBenchmark`.
+
+NOTE: Benchmark methods need to be annotated by `@Benchmark` for JMH to detect them.
+
+==== Running benchmarks from Maven
+
+To easily run benchmrks from Maven, a Maven profile to run the benchmarks has been created
+and is available starting Plugin-POM version 3.45. You can then run your benchmarks from the
+command line using `mvn test -Dbenchmark`. To make it easier for plugins hosted on ci.jenkins.io,
+a build step has been added in Jenkins pipeline library which automatically runs benchmarks
+and optionally archives the benchmark results.
+
+=== Sample benchmarks
+
+First, you need to have a benchmark runner that contains a `@Test` so it can run
+like a JUnit test:
+
+[source, java]
+----
+public class BenchmarkRunner {
+    @Test
+    public void runJmhBenchmarks() throws Exception {
+        ChainedOptionsBuilder options = new OptionsBuilder()
+                .mode(Mode.AverageTime)
+                .forks(2)
+                .result("jmh-report.json");
+
+        // Automatically detect benchmark classes annotated with @JmhBenchmark
+        BenchmarkFinder.findBenchmarks(options);
+        new Runner(options.build()).run();
+    }
+}
+----
+
+This allows us to automatically find benchmarks and run them through a JUnit test.
+Since `BenchmarkRunner` class name does not qualify as a test according to Maven surefire plugin's
+naming conventions, the benchmarks will not interfere with your JUnit tests.
+Now, you can write your benchmarks:
+
+**Without any special setup**:
+
+[source,java]
+----
+@JmhBenchmark
+public class JmhStateBenchmark {
+    public static class MyState extends JmhBenchmarkState {
+    }
+
+    @Benchmark
+    public void benchmark(MyState state) {
+        // benchmark code goes here
+        state.getJenkins().setSystemMessage("Hello world");
+    }
+}
+----
+
+// TODO: Add docs for configuration as code after configuration-as-code#921 is merged
```

Comment:
Maybe all configuration-as-code logic could go to a separate section. It is an interesting content, and it worth highligting it. You can start drafting the doc IMHO. Even if the delivery gets delayed, in the worst case we could just move it to a separate blogpost.